deployment/datastore/memory_address_reference_store_test.go

```diff
@@ -0,0 +1,807 @@
+			givenState: []AddressReferenceRecord{
+				recordOne,
+				recordTwo,
+			},
+			giveKey:        recordTwo.Key(),
+			expectedRecord: recordTwo,
+			expectedError:  nil,
+		},
+		{
+			name:           "error: record not found",
+			givenState:     []AddressReferenceRecord{},
+			giveKey:        recordTwo.Key(),
+			expectedRecord: AddressReferenceRecord{},
+			expectedError:  ErrAddressReferenceRecordNotFound,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			store := InMemoryAddressReferenceStore{records: tt.givenState}
+			record, err := store.Get(tt.giveKey)
+
+			if tt.expectedError != nil {
+				require.Error(t, err)
+				assert.Equal(t, tt.expectedError, err)
+			} else {
+				require.NoError(t, err)
+				assert.Equal(t, tt.expectedRecord, record)
+			}
+		})
+	}
+}
+
+func TestInMemoryAddressReferenceStore_Filter(t *testing.T) {
+	var (
+		recordOne = AddressReferenceRecord{
+			Address:   "0x2324224",
+			Chain:     1,
+			Type:      "type1",
+			Version:   semver.MustParse("0.5.0"),
+			Qualifier: "qual1",
+			Labels: deployment.NewLabelSet(
+				"label1", "label2", "label3",
+			),
+		}
+
+		recordTwo = AddressReferenceRecord{
+			Address:   "0x2324224",
+			Chain:     2,
+			Type:      "typeX",
+			Version:   semver.MustParse("0.5.0"),
+			Qualifier: "qual1",
+			Labels: deployment.NewLabelSet(
+				"label13", "label23", "label33",
+			),
+		}
+
+		recordThree = AddressReferenceRecord{
+			Address:   "0x2324224",
+			Chain:     3,
+			Type:      "typeZ",
+			Version:   semver.MustParse("0.5.0"),
+			Qualifier: "qual1",
+			Labels: deployment.NewLabelSet(
+				"label13", "label23", "label33",
+			),
+		}
+	)
+
+	tests := []struct {
+		name           string
+		givenState     []AddressReferenceRecord
+		giveFilters    []FilterFunc[AddressReferenceKey, AddressReferenceRecord]
+		expectedResult []AddressReferenceRecord
+	}{{
+		name: "success: no filters returns all records",
+		givenState: []AddressReferenceRecord{
+			recordOne,
+			recordTwo,
+			recordThree,
+		},
+		giveFilters:    []FilterFunc[AddressReferenceKey, AddressReferenceRecord]{},
+		expectedResult: []AddressReferenceRecord{recordOne, recordTwo, recordThree},
+	},
+		{
+			name: "success: returns record with given chain and type",
+			givenState: []AddressReferenceRecord{
+				recordOne,
+				recordTwo,
+				recordThree,
+			},
+			giveFilters: []FilterFunc[AddressReferenceKey, AddressReferenceRecord]{
+				AddressReferenceRecordByChain(2),
+				AddressReferenceRecordByType("typeX"),
+			},
+			expectedResult: []AddressReferenceRecord{recordTwo},
+		},
+		{
+			name: "success: returns no record with given chain and type",
+			givenState: []AddressReferenceRecord{
+				recordOne,
+				recordTwo,
+				recordThree,
+			},
+			giveFilters: []FilterFunc[AddressReferenceKey, AddressReferenceRecord]{
+				AddressReferenceRecordByChain(4),
+				AddressReferenceRecordByType("typeX"),
+			},
+			expectedResult: []AddressReferenceRecord(nil),
```

Comment:
i would expect this instead of nil?
```suggestion
			expectedResult: []AddressReferenceRecord{},
```

Reply:
Good point! Switching var filtered []AddressRef to filtered := []AddressRef{} inside filters should fix that.